clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java

```diff
@@ -151,12 +156,12 @@ public static Builder forMagic(byte magic,
             return new Builder(minVersion, maxVersion, acks, timeout, partitionRecords, transactionalId);
         }
 
-        private Builder(short minVersion,
-                        short maxVersion,
-                        short acks,
-                        int timeout,
-                        Map<TopicPartition, MemoryRecords> partitionRecords,
-                        String transactionalId) {
+        public Builder(short minVersion,
```

Comment:
Do we need it? The instantiation of `MemoryRecordsBuilder` with zstd and magic < 2 is already disallowed so the user can't create `ProduceRequest` instance also.

Reply:
I'm not sure I follow. Produce request v6 supports magic=2, so what prevents a user from sending a zstd batch to the broker with v6?

Reply:
Since `CompressionType` is configured on `RecordBatch` instances in `MemoryRecords,` I think we don't need to add a new, duplicate parameter. How about checking whether `MemoryRecords` instances include the zstd compressed `RecordBatch` instances? You know, the compression type may differ between topic partitions included in `ProduceRequest.`

Reply:
Yeah, I think that works also. Actually, now that you mention it, another place we could do this validation is `ProduceRequest.validateRecords`.

Reply:
This approach seems like the most reasonable. :+1: for `ProduceRequest.validateRecords`.